File: backend/GameApp.Domain/ValueObjects/Characters/Character.cs

using GameApp.Domain.ValueObjects.Items;

namespace GameApp.Domain.ValueObjects.Characters;

// Abstract class for Character value object
public abstract class Character
{
    public CharacterName Name { get; }
    public int MaxHealthPoints { get; }
    public int MaxFoodPoints { get; }
    public int MaxInventorySlots { get; }
    public int StartingMoney { get; }


    // current Character atributes
    public int CurrentHealthPoints { get; }
    public int CurrentFoodPoints { get; }
    public int CurrentMoney { get; }
    public IReadOnlyList<Item> InventoryList { get; }


    // Constructor
    protected Character(CharacterName name, int maxHealthPoints, int maxFoodPoints, int maxInventorySlots, int startingMoney)
    {
        Name = name;
        MaxHealthPoints = maxHealthPoints;
        MaxFoodPoints = maxFoodPoints;
        MaxInventorySlots = maxInventorySlots;
        StartingMoney = startingMoney;

        CurrentHealthPoints = MaxHealthPoints;
        CurrentFoodPoints = MaxFoodPoints;
        CurrentMoney = StartingMoney;
        InventoryList = new List<Item>();
    }

    // Restore Constructor
    protected Character(CharacterName name, int maxHealthPoints, int maxFoodPoints, int maxInventorySlots, int startingMoney,
                        int currentHealthPoints, int currentFoodPoints, int currentMoney, List<Item> inventoryList)
    {
        Name = name;
        MaxHealthPoints = maxHealthPoints;
        MaxFoodPoints = maxFoodPoints;
        MaxInventorySlots = maxInventorySlots;
        StartingMoney = startingMoney;

        CurrentHealthPoints = currentHealthPoints;
        CurrentFoodPoints = currentFoodPoints;
        CurrentMoney = currentMoney;
        InventoryList = inventoryList;
    }

    // Clone method for inmutable setters
    protected abstract Character CloneWith(int currentHealthPoints, int currentFoodPoints, int currentMoney, IReadOnlyList<Item> inventoryList);

    // getters
    public CharacterName GetName()
    {
        return Name;
    }

    public int GetMaxHealthPoints()
    {
        return MaxHealthPoints;
    }

    public int GetMaxFoodPoints()
    {
        return MaxFoodPoints;
    }

    public int GetMaxInventorySlots()
    {
        return MaxInventorySlots;
    }

    public int GetStartingMoney()
    {
        return StartingMoney;
    }

    public int GetCurrentHealthPoints()
    {
        return CurrentHealthPoints;
    }

    public int GetCurrentFoodPoints()
    {
        return CurrentFoodPoints;
    }

    public int GetCurrentMoney()
    {
        return CurrentMoney;
    }

    public IReadOnlyList<Item> GetInventoryList()
    {
        return InventoryList;
    }

    // setters
    public Character AddItemInventory(Item newItem)
    {
        if (InventoryList.Count >= MaxInventorySlots)
            return this;

        var newInventory = new List<Item>(InventoryList) { newItem };
        return CloneWith(CurrentHealthPoints, CurrentFoodPoints, CurrentMoney, newInventory);
    }

    public Character RemoveItemInventory(int index)
    {
        if (index < 0 || index >= InventoryList.Count)
            return this;

        var newInventory = new List<Item>(InventoryList);
        newInventory.RemoveAt(index);
        return CloneWith(CurrentHealthPoints, CurrentFoodPoints, CurrentMoney, newInventory);
    }

    public Character ReceiveDamage(int damage)
    {
        int newHealth = Math.Max(0, CurrentHealthPoints - damage);
        return CloneWith(newHealth, CurrentFoodPoints, CurrentMoney, InventoryList);
    }

    public Character Heal(int amount)
    {
        int newHealth = Math.Min(MaxHealthPoints, CurrentHealthPoints + amount);
        return CloneWith(newHealth, CurrentFoodPoints, CurrentMoney, InventoryList);
    }

    public Character Eat(int amount)
    {
        int newFood = Math.Min(MaxFoodPoints, CurrentFoodPoints + amount);
        return CloneWith(CurrentHealthPoints, newFood, CurrentMoney, InventoryList);
    }

    public Character GetHungry(int amount)
    {
        int newFood = Math.Max(0, CurrentFoodPoints - amount);
        return CloneWith(CurrentHealthPoints, newFood, CurrentMoney, InventoryList);
    }

    public Character EarnMoney(int amount)
    {
        return CloneWith(CurrentHealthPoints, CurrentFoodPoints, CurrentMoney + amount, InventoryList);
    }

    public Character SpendMoney(int amount)
    {
        int newMoney = Math.Max(0, CurrentMoney - amount);
        return CloneWith(CurrentHealthPoints, CurrentFoodPoints, newMoney, InventoryList);
    }

}

File: frontend/gameapp-frontend-js/src/domain/value-objects/characters/character.js
import { Item } from "../../entities/items/item";
import {CharacterName} from './character-name';


// Abstract base class Character
export class Character {
  constructor(name, maxHealthPoints, maxFoodPoints, maxInventorySlots, startingMoney,
              currentHealthPoints = null, currentFoodPoints = null, currentMoney = null, inventoryList = null) {
    if (!(name instanceof CharacterName)) throw new TypeError("name must be CharacterName");

    this._name = name;
    this._maxHealthPoints = maxHealthPoints;
    this._maxFoodPoints = maxFoodPoints;
    this._maxInventorySlots = maxInventorySlots;
    this._startingMoney = startingMoney;

    this._currentHealthPoints = currentHealthPoints ?? maxHealthPoints;
    this._currentFoodPoints = currentFoodPoints ?? maxFoodPoints;
    this._currentMoney = currentMoney ?? startingMoney;
    this._inventoryList = inventoryList ?? [];

    this.#validateItemList(this._inventoryList, "_inventoryList");
  }

  #validateItemList(list, paramName) {
          if (!Array.isArray(list)) throw new TypeError(`${paramName} must be an array`);
          for (const item of list) {
              if (!(item instanceof Item)) {
                  throw new TypeError(`All elements of ${paramName} must be instances of Item or its subclasses`);
              }
          }
      }

  // getters
  get name() { return this._name; }
  get maxHealthPoints() { return this._maxHealthPoints; }
  get maxFoodPoints() { return this._maxFoodPoints; }
  get maxInventorySlots() { return this._maxInventorySlots; }
  get startingMoney() { return this._startingMoney; }
  get currentHealthPoints() { return this._currentHealthPoints; }
  get currentFoodPoints() { return this._currentFoodPoints; }
  get currentMoney() { return this._currentMoney; }
  get inventoryList() { return this._inventoryList; }

  // Abstract clone method
  cloneWith(currentHealthPoints, currentFoodPoints, currentMoney, inventoryList) {
    throw new Error("Method 'cloneWith' must be implemented in subclass");
  }

  // setters
  addItemInventory(newItem) {
    if (this._inventoryList.length >= this._maxInventorySlots) return this;

    const newInventory = [...this._inventoryList, newItem];
    return this.cloneWith(this._currentHealthPoints, this._currentFoodPoints, this._currentMoney, newInventory);
  }

  removeItemInventory(index) {
    if (index < 0 || index >= this._inventoryList.length) return this;

    const newInventory = [...this._inventoryList];
    newInventory.splice(index, 1);
    return this.cloneWith(this._currentHealthPoints, this._currentFoodPoints, this._currentMoney, newInventory);
  }

  receiveDamage(damage) {
    const newHealth = Math.max(0, this._currentHealthPoints - damage);
    return this.cloneWith(newHealth, this._currentFoodPoints, this._currentMoney, this._inventoryList);
  }

  heal(amount) {
    const newHealth = Math.min(this._maxHealthPoints, this._currentHealthPoints + amount);
    return this.cloneWith(newHealth, this._currentFoodPoints, this._currentMoney, this._inventoryList);
  }

  eat(amount) {
    const newFood = Math.min(this._maxFoodPoints, this._currentFoodPoints + amount);
    return this.cloneWith(this._currentHealthPoints, newFood, this._currentMoney, this._inventoryList);
  }

  getHungry(amount) {
    const newFood = Math.max(0, this._currentFoodPoints - amount);
    return this.cloneWith(this._currentHealthPoints, newFood, this._currentMoney, this._inventoryList);
  }

  earnMoney(amount) {
    return this.cloneWith(this._currentHealthPoints, this._currentFoodPoints, this._currentMoney + amount, this._inventoryList);
  }

  spendMoney(amount) {
    const newMoney = Math.max(0, this._currentMoney - amount);
    return this.cloneWith(this._currentHealthPoints, this._currentFoodPoints, newMoney, this._inventoryList);
  }
}